A34-edf-weighted-round-robin.md

```diff
@@ -0,0 +1,111 @@
+
+#### Service Config
+The service config for the `edf_weighted_round_robin` LB policy is an empty proto message
+```
+{
+  load_balancing_config: { edf_weighted_round_robin: {}}
+}
+```
+
+### Handling per endpoint `load_balancing_weight` from `ClusterLoadAssignment` response
+
+This part extends the behavior of EDS section in [A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md). Instead of discarding the per endpoint `load_balancing_weight`, we want to add it to per-address attibute and pass it along to `lb_policy`.
+
+#### `lb_policy` for per endpoint `load_balancing_weight` from `ClusterLoadAssignment`
+When the `lb_policy` field in CDS response is `ROUND_ROBIN`, we use `edf_weighted_round_robin` as the `lb_policy`.
+
+As of today, we only accept `ROUND_ROBIN` as `lb_policy` in CDS response per [A27: xDS-Based Global Load Balancing](https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md). Therefore, `edf_weighted_round_robin` will always be used.
+
+#### On update of `ClusterLoadAssignment`
+When an EDS update is received, an update will be sent to the `lb_policy`. The `lb_policy` will create a new picker. This is slightly Different from [Envoy](https://github.com/envoyproxy/envoy/blob/51551ae944c642e6fc61563cbea8653087e70f1f/source/common/upstream/load_balancer_impl.cc#L733-L737). We'd like to udpate EDF priority queue so that new weights applied immediately even endpoints list is not changed.
+
+#### NOTE
+- `edf_weighted_round_robin` should always be updated to the lastest `ClusterLoadAssignment`. It's xDS server's responsibility to maintain consistency.
+
+## Rationale
+
+Several applications can be built upon this feature, e.g. utilization load balancing, blackhole erroring endpoints, load testing,... etc.
+
+The reason to refresh EDF picker even there is only weight change on some endpoints which is different from envoy is because we'd like real time traffic shift for use cases like load testing, blackhole erroring endpoints.
+
+The reasons to introduce a new algorithm instead of re-using the same algorithm of `weighted_target` policy are
+- [EDF](https://en.wikipedia.org/wiki/Earliest_deadline_first_scheduling) maintains FIFO order for endpoints with same weight which is easier to reason.
+- We want to be consistent with the behavior of Envoy on how lb_policy picks the backend for sending the request. However, there is a difference between `edf_weighted_round_robin` and `edf_scheduler` of Envoy. `edf_weighted_round_robin` actively monitors the connectivity of each subchannel but `edf_scheduler` of Envoy does not.
+
+## Implementation
+
+N/A
+
+## Open issues (if applicable)
+* How to desync to avoid all clients do synchronized pick?
```

Comment:
I don't understand what this is referring to.  Can you explain the problem here?

Reply:
updated. Please let me know it's still not clear

Reply:
Okay, I understand the problem.  I see two possible solutions here:

- The simple approach is, as you suggest, randomize the starting point in the list whenever we create a new picker, just like `round_robin` does.  This will avoid having all clients hammer the same endpoints at the same time (and also, keep in mind that the control plane could actually send different results to different clients -- each client could see different endpoints or the same endpoints in a different order or with different weights).  But it will disrupt the expected scheduling whenever the picker changes.

- A more complicated solution would be to maintain the current scheduler state in the LB policy and use a mutex to synchronize access to it between the picker and the LB policy.  This is more complicated to implement, and the synchronization imposes a performance penalty (because you have to acquire the mutex for every pick), but it should work.